journals/week5.md

```diff
@@ -0,0 +1,186 @@
+
+```bash
+$ gcc -shared -I$HOME/flux-core -I$HOME/flux-core/src/include -fPIC \
+  -o state-counts.so ../fripp.h ../fripp.c state-counts.c
+```
+
+Then from `flux-core` in a running instance, the plugin can be loaded (assuming it's in `../fripp/code/jobtap-plugins/`) by:
+
+```bash
+$ flux jobtap load $(pwd)/../fripp/code/jobtap-plugins/state-counts.so
+```
+
+After loading the plugin I ran 10,000 jobs using the throughput test in a single node instance.
+
+**Sched vs Inactive**
+
+![job states sched vs inactive](../img/gr_flux_js_is.png)
+
+**Cleanup vs Run**
+
+![job states cleanup vs run](../img/gr_flux_js_cr.png)
+
+
+**Priority vs Depend**
+
+![job states priority vs depend](../img/gr_flux_js_pd.png)
+
+**New**
+
+![job states  new](../img/gr_flux_js_n.png)
+
+The `sched` and `inactive` states have exact opposite trends which makes sense. In the beginning there should be 10,000 jobs scheduled and 0 inactive, and gradually the scheduled jobs should get run and cleaned up resulting in 10,000 inactive and 0 left scheduled. The `run` and `cleanup` states also seemed to have opposite trends, and it makes sense that if lots of jobs are running that not many jobs would be being cleaned up. Near the end there is a large spike in cleanups (and respective dip in running jobs) which is quite interesting. `Priority` and `depend` are two states that in this test, and every subsequent one, were always 0. If i print the states from within the plugin's `state_cb`, on a single job it produces the following output:
+
+```bash
+$ ./src/test/throughput.py -n1 -x
+state=DEPEND prev_state=NEW
+state=PRIORITY prev_state=DEPEND
+state=SCHED prev_state=PRIORITY
+state=RUN prev_state=SCHED
+state=CLEANUP prev_state=RUN
+state=INACTIVE prev_state=CLEANUP
+```
+
+And it's clear to see that the job does go thru both the `priority` and `depend` states. My assumption is that the resolution on some combination of the sending of the metrics and aggregating the metrics is not fine enough to mark these transitions as they are very quick. Lastly, why is the new state is getting to around -3.7 million?? Well that is because of two things. First the reason it's so large is because of the way that gauges in brubeck/statsd function. They take and arbitrary *positive* value or a signed value to increment the previous value. The reason that's effecting us is because of the topic string that the `state_cb` was registered under. `job.state.*` will never enter with a current state of `new`, so each time a new job is introduced the plugin catches it on the way out of the `new` state causing the count to be -10,000. Putting those two pieces together, every second when counts are being sent, the `flux.job-states.new` tag is being decremented by 10,000 instead of being set to -10,000. To handle this, I registered the same `state_cb` under the `job.new` topic string which will catch the jobs on the way in and *hopefully* keep the number non-negative.
```

Comment:
Good analysis here!

As jobs are ingested in the job manager, they immediately enter the DEPEND state. If no plugins have emitted `dependency-add` events (i.e. there are no dependencies) then the job will _immediately_ have a `depend` event posted and will proceed to the PRIORITY state. Unless there is a non-default priority plugin loaded, the job will then _immediately_ have a `priority` event posted and will proceed to the SCHED state.

Therefore, in the case of the tests being run here, jobs spend almost no time in the DEPEND and PRIORITY states, but instead proceed immediately to the SCHED state.